pynapple/process/decoding.py

```diff
@@ -134,64 +134,81 @@ def decode_2d(tuning_curves, group, feature, ep, bin_size, xy):
+
+    if not np.all(np.array(list(tuning_curves.keys())) == np.array(newgroup.keys())):
+        raise RuntimeError("Difference indexes for tuning curves and group keys")	
 
 	# Bin spikes
-	count = newgroup.count(bin_size, ep, 's')
+    if type(newgroup) is not nap.TsdFrame:
+        count = newgroup.count(bin_size, ep, 's')
+    else:
+        #Spikes already "binned" with continuous TsdFrame input
+        count = newgroup
 
 	# Occupancy
-	binsxy = []
-	for i in range(len(xy)):
-		diff = np.diff(xy[i])
-		bins = xy[i][:-1] - diff/2
-		bins = np.hstack((bins, [bins[-1]+diff[-1],bins[-1]+2*diff[-1]])) # assuming the size of the last 2 bins is equal
-		binsxy.append(bins)
-
-	occupancy, _, _ = np.histogram2d(
-		feature.iloc[:,0],
-		feature.iloc[:,1],
-		[binsxy[0], binsxy[1]])
-	occupancy = occupancy.flatten()
+    if feature is None:
+        #For conditions (e.g. sleep) where you don't have, or don't want to 
+        #bias by behavioral occupancy for decoding
+        occupancy = np.ones(1)
+        #TODO: Assuming features are x and y.... this is hacky
+        feature = nap.TsdFrame(1,columns = ('x','y'))
+    else:
+        binsxy = []
+        for i in range(len(xy)):
+            diff = np.diff(xy[i])
+            bins = xy[i][:-1] - diff/2
+            bins = np.hstack((bins, [bins[-1]+diff[-1],bins[-1]+2*diff[-1]])) # assuming the size of the last 2 bins is equal
+            binsxy.append(bins)
+    
+        occupancy, _, _ = np.histogram2d(
+    		feature.iloc[:,0],
+    		feature.iloc[:,1],
+    		[binsxy[0], binsxy[1]])
+        occupancy = occupancy.flatten()
 
 	# Transforming to pure numpy array
-	tc = np.array([tuning_curves[i] for i in tuning_curves.keys()])
-	tc = tc.reshape(tc.shape[0], np.prod(tc.shape[1:]))
-	tc = tc.T
+    tc = np.array([tuning_curves[i] for i in tuning_curves.keys()])
+    tc = tc.reshape(tc.shape[0], np.prod(tc.shape[1:]))
+    tc = tc.T
 
-	ct = count.values
+    ct = count.values
 
-	p1 = np.exp(-bin_size*np.nansum(tc, 1))
-	p2 = occupancy/occupancy.sum()
+    p1 = np.exp(-bin_size*np.nansum(tc, 1))
+    p2 = occupancy/occupancy.sum()
 
-	ct2 = np.tile(ct[:,np.newaxis,:], (1,tc.shape[0],1))
+    ct2 = np.tile(ct[:,np.newaxis,:], (1,tc.shape[0],1))
 
-	p3 = np.nanprod(tc**ct2, -1)
+    p3 = np.nanprod(tc**ct2, -1)
 
-	p = p1 * p2 * p3
-	p = p / p.sum(1)[:,np.newaxis]
+    p = p1 * p2 * p3
+    p = p / p.sum(1)[:,np.newaxis]
 
-	idxmax = np.argmax(p, 1)
+    idxmax = np.argmax(p, 1)
 
-	p = p.reshape(p.shape[0], len(xy[0]), len(xy[1]))
+    p = p.reshape(p.shape[0], len(xy[0]), len(xy[1]))
 
-	idxmax2d = np.unravel_index(idxmax, (len(xy[0]), len(xy[1])))
+    idxmax2d = np.unravel_index(idxmax, (len(xy[0]), len(xy[1])))
 
-	decoded = nap.TsdFrame(
+    decoded = nap.TsdFrame(
 		t = count.index.values,
 		d = np.vstack((xy[0][idxmax2d[0]], xy[1][idxmax2d[1]])).T,
 		time_support = ep,
 		columns=feature.columns
 		)
 
-	return decoded, p
\ No newline at end of file
+    return decoded, p
```

Comment:
I think it should work but can you provide with a minimal working example with toy data in your PR that can generate some figures to have a full check.

Reply:
I don't have any continuous neural data, but could provide an example with neural network data that makes (albeit still poorly) place fields and decoding? 😅